doc/man3/strftime.3
.TH STRFTIME 3
.SH NAME
strftime - formats the broken-down time according to specified format
.SH SYNOPSIS
#include <time.h>

.nf
size_t strftime(char * restrict s,
                size_t maxsize,
                const char * restrict format,
                const struct tm * restrict timeptr);
.fi
.SH DESCRIPTION
The
.BR strftime ()
function places characters into the array pointed to by
.I s
as controlled by the string pointed to by
.IR format .

The
.I format
shall be a multibyte character sequence,
beginning and ending in its initial shift state.
The
.I format
shall consist of zero or more conversion specifiers and
ordinary characters.
A conversion specifier consists of a % character,
possibly followed by an E or O modifier character,
followed by a character that determines the behavior of the conversion
specifier.
All ordinary characters
(including the terminating null character) are copied unchanged into the
array.
No more than
.I maxsize
characters are placed into the array.

The format specification is a null-terminated string and
shall contain special character sequences, called conversion specifiers,
which get replaced with the appropriate characters as described in the
following list:
.TP
.B %a
The abbreviated name of the day of the week according to the current
locale.
.TP
.B %A
THe full name of the day of the week according to the current locale.
.TP
.B %b
The abbreviated name of the month according to the current locale.
.TP
.B %B
The full name of the month according to the current locale.
.TP
.B %c
The preferred date and time representation for the current locale.
.TP
.B %C
The decimal number representing the year divided by 100 and truncated
to an integer (century number).
.TP
.B %d
The day of the month as a decimal number between 01 and 31.
.TP
.B %D
Equivalent to "%m/%d/%y"
.TP
.B %e
The day of the month as a decimal number between 0 and 31;
a single digit is preceded by a space.
.TP
.B %F
Equivalent to "%y-%m-%d"
(the ISO 8601 date format).
.TP
.B %G
The ISO 8601 week-based-year with century as a decimal number.
The 4-digit year corresponding to the ISO week number. (Calculated
from
.BR tm_year ,
.BR tm_yday ,
and
.BR tm_wday )
.TP
.B %g
Like
.B %G
but without century, that is, with a 2-digit year (00-99).
(Calculated from
.BR tm_year ,
.BR tm_yday ,
and
.BR tm_wday )
.TP
.B %h
Equivalent to
.BR %b .
(Calculated from
.BR tm_mon )
.TP
.B %H
The hour as a decimal number using a 24-hour clock between
00 and 23.
(Calculated from
.BR tm_hour )
.TP
.B %I
The hour as a decimal number using a 12-hour clock between 01 and 12
(Calculated from
.BR tm_hour )
.TP
.B %j
The day of the year as a decimal number between 001 and 366.
(Calculated from
.BR tm_yday )
.TP
.B %m
The month as a decimal number between 01 and 12
(Calculated from
.BR tm_mon )
.TP
.B %M
The minute as a decimal number between 00 and 59
.TP
.B %n
A newline character.
.TP
.B %p
Either "AM" or "PM" according to the given time value.
(Calculated from
.BR tm_hour )
.B %r
The locale's 12-hour clock time.
.TP
.B %R
The time in 24-hour notation (%H:%M).
.TP
.B %S
The second as a decimal number between 00 and 60.
(Calculated from
.BR tm_sec ).
.TP
.B %t
A tab character.
.TP
.B %T
The time in 24-hour notation (%H:%M:%S).
.TP
.B %u
The day of the week (ISO 8601) as a decimal between 1 and 7,
Monday being 1.
(Calculated from
.BR tm_wday )
.TP
.B %U
The week number of the current year as a decimal number
between 00 and 53,
starting with the first Sunday as the first day of week 01. (Calculated from
.BR tm_year ,
.BR tm_wday ,
.BR tm_yday )
.TP
.B %V
The ISO 8601 week (see NOTES) number of the current year as a decimal number
between 01 and 53,
where week 1 is the first week that has at least 4 days in the new year.
.TP
.B %w
The day of the week as a decimal between 0 and 6,
Sunday being 0.
.TP
.B %W
The week number of the current year as a decimal number
between 00 and 53,
starting with the first Monday as the first day of week 01.
(Calculated from
.B tm_yday
and
.BR tm_wday ).
.TP
.B %x
The preferred date representation for the current locale without the
time.
.TP
.B %X
The preferred time representation for the current locale without the
date.
.TP
.B %y
The year as a decimal number without a century between 00 and 99.
.TP
.B %Y
The year as a decimal number including the century.
.TP
.B %z
The +hhmm or -hhmm numeric timezone
(that is, the hour and minute offset from UTC in ISO 8601 format).
.TP
.B %Z
The timezone name or abbreviation.
.TP
.B %%
This is replaced with %.
.P
Some conversion specifiers can be modified by the inclusion of an
.B E
or
.B O
modifier character to indicate an alternative format or specification.
If the alternative format or character doesn't exist for the current locale,
the modifier is ignored.

The C Standard mentions the following specifiers:
.BR %Ec ,
.BR %EC ,
.BR %Ex ,
.BR %EX ,
.BR %Ey ,
.BR %EY ,
.BR %Od ,
.BR %Oe ,
.BR %OH ,
.BR %OI ,
.BR %Om ,
.BR %OM ,
.BR %OS ,
.BR %Ou ,
.BR %OU ,
.BR %OV ,
.BR %Ow ,
.BR %OW ,
.BR %Oy ,
where the effect of the
.B O
modifier is to use
alternative numeric symbols, and that of the
.B E
modifier is to use a locale-dependent alternative representation.

In the "C" locale, the E and O modifiers are ignored and the
replacement strings for the following specifiers are:
.TP
.B %a
the first three characters of %A
.TP
.B %A
one of Sunday ,Monday, ..., Saturday
.TP
.B %b
the first three characters of %B
.TP
.B %B
one of January, February, ..., December
.TP
.B %c
equivalent to %a %b %e %T %Y
.TP
.B %p
one of AM or PM
.TP
.B %r
equivalent to %I:%M:%S %p
.TP
.B %x
equivalent to %m/%d/%y
.TP
.B %X
equivalent to %T
.TP
.B %Z
The timezone name or abbreviation.
.SH RETURN VALUE
If the total number of resulting characters
including the terminating null character doesn't exceed
.IR maxsize ,
then
.I strftime
function returns the number of characters
placed into the array pointed to by
.I s
not including the terminating null character. Otherwise, zero is
returned and the contents of the array are indeterminate.
.SH NOTES
In the ISO 8601 week-based year, weeks begin on a Monday and week 1
of the year is the week that includes January 4th, which also includes
the first Thursday of the year, and is also the first week that
contains at least four days in the year. As mentioned above,
.B %g, %G,
and
.B %V
are dependent on ISO 8601 week-based year, and the following examples
are provided for illustrative purposes:

If the first Monday of January is the 2nd, 3rd, or 4th,
the preceding days are part of the last week of the preceding year;
thus, for Saturday 2nd January 1999,
.B %G
is replaced by 1998 and
.B %V
is replaced with 53. Similarly,
if December 29th, 30th or 31st is a Monday,
it and following days are part of week 1 and hence,
for Tuesday 30th December 1997,
.B %G
is replaced by 1998 and
.B %V
is replaced with 01.
.SH STANDARDS
ISO/IEC 9899:1999 Section 7.23.3.5 Paragraph 1,2,3,4,5,6,7
.SH SEE ALSO
.BR time.h (3)


src/libc/time/strftime.c
#include <string.h>
#include <time.h>

#include "../libc.h"

#undef strftime

static char *days[] = {
	"Sunday",   "Monday", "Tuesday",  "Wednesday",
	"Thursday", "Friday", "Saturday",
};

static char *months[] = {
	"January",   "February", "March",    "April",
	"May",       "June",     "July",     "August",
	"September", "October",  "November", "December"
};

static char *am_pm[] = {"AM", "PM"};

static int
first(int day, int year)
{
	int ny;

	ny = _newyear(year);
	if (ny == day)
		return 0;
	if (day - ny < 0)
		return 7 - (ny - day);
	return day - ny;
}

static int
weeknum(struct tm *tm, int day)
{
	int fday, val;

	fday = first(day, tm->tm_year);
	if (tm->tm_yday < fday) {
		val = 0;
	} else {
		val = tm->tm_yday - fday;
		val /= 7;
		val++;
	}
	return val;
}

static int
isoyear(struct tm *tm)
{
	int monday;

	if (tm->tm_yday < 7) {
		monday = first(THU, tm->tm_year) - 3;
		if (tm->tm_yday < monday)
			return tm->tm_year - 1;
	} else if (tm->tm_yday > 357) {
		monday = first(THU, tm->tm_year + 1) - 3;
		if (tm->tm_mday >= (31 + monday))
			return tm->tm_year + 1;
	}
	return tm->tm_year;
}

static int
isoweek(struct tm *tm)
{
	int year, monday, yday, val;

	year = isoyear(tm);
	monday = first(THU, year) - 3;
	yday = tm->tm_yday;
	if (year > tm->tm_year)
		yday = tm->tm_mday - 31 + monday;
	else if (year < tm->tm_year)
		yday = _daysyear(year) + yday;
	val = yday - monday;
	val /= 7;
	val++;
	return val;
}

static int
isoday(struct tm *tm)
{
	if (tm->tm_wday == 0)
		return 7;
	return tm->tm_wday;
}

static size_t
sval(char *s, size_t siz, char **strs, int abrev, int idx, int max)
{
	char *str;
	size_t len;

	if (idx < 0 && idx >= max)
		goto wrong;

	str = strs[idx];
	len = (!abrev) ? strlen(str) : 3;
	if (len > siz)
		goto wrong;

	memcpy(s, str, len);
	return len;

wrong:
	*s = '?';
	return 1;
}

static size_t
dval(char *s, size_t siz, int prec, int fill, int val)
{
	char *t;
	int n;
	static char digits[] = "0123456789";

	if (prec > siz || val < 0) {
		*s = '?';
		return 1;
	}

	n = prec;
	do {
		s[--n] = digits[val % 10];
		val /= 10;
	} while (n > 0 && val > 0);

	while (n > 0)
		s[--n] = fill;

	return prec;
}

static size_t
timezone(char *s, size_t prec, const struct tm * restrict tm)
{
	long off = tm->tm_gmtoff;

        if (prec < 5) {
                *s = '?';
                return 1;
        }

	if (off >= 0) {
		*s++ = '+';
	} else {
		*s++ = '-';
		off = -off;
	}

	dval(s, 2, 2, '0', off / 3600);
	dval(s, 2, 2, '0', (off % 3600) / 60);

	return 5;
}

size_t
strftime(char *restrict s, size_t maxsize,
	 const char *restrict format, const struct tm *restrict timeptr)
{
	int ch, abrev, val, fill, width;
	size_t n, inc;
	char *tfmt;

	for (n = --maxsize; (ch = *format++) && n > 0; s += inc, n -= inc) {
		if (ch != '%') {
			*s = ch;
			inc = 1;
			continue;
		}

		abrev = 0;
		fill = '0';
		width = 2;

		if (*format == 'E' || *format == 'O')
			format++;

		switch (*format++) {
		case 'a':
			abrev = 1;
		case 'A':
			inc = sval(s, n, days, abrev, timeptr->tm_wday, 7);
			break;
		case 'h':
		case 'b':
			abrev = 1;
		case 'B':
			inc = sval(s, n, months, abrev, timeptr->tm_mon, 12);
			break;
		case 'c':
			tfmt = "%a %b %e %T %Y";
			goto recursive;
		case 'C':
			val = (timeptr->tm_year + 1900) / 100;
			goto number;
		case 'd':
			val = timeptr->tm_mday;
			goto number;
		case 'D':
			tfmt = "%m/%d/%y";
			goto recursive;
		case 'e':
			fill = ' ';
			val = timeptr->tm_mday;
			goto number;
		case 'F':
			tfmt = "%Y-%m-%d";
			goto recursive;
		case 'g':
			val = isoyear(timeptr);
			goto number;
		case 'G':
			val = isoyear(timeptr);
			val += 1900;
			width = 4;
			goto number;
		case 'H':
			val = timeptr->tm_hour;
			goto number;
		case 'I':
			val = timeptr->tm_hour;
			if (val == 0)
				val = 12;
			if (val > 12)
				val -= 12;
			goto number;
		case 'j':
			width = 3;
			val = timeptr->tm_yday+1;
			goto number;
		case 'm':
			val = timeptr->tm_mon+1;
			goto number;
		case 'M':
			val = timeptr->tm_min;
			goto number;
		case 'n':
			val = '\n';
			goto character;
		case 'p':
			inc = sval(s, n, am_pm, 0, timeptr->tm_hour > 12, 2);
			break;
		case 'r':
			tfmt = "%I:%M:%S %p";
			goto recursive;
		case 'R':
			tfmt = "%H:%M";
			goto recursive;
		case 'S':
			val = timeptr->tm_sec;
			goto number;
		case 't':
			val = '\t';
			goto character;
		case 'u':
			width = 1;
			val = isoday(timeptr);
			goto number;
		case 'U':
			val = weeknum(timeptr, SUN);
			goto number;
		case 'V':
			val = isoweek(timeptr);
			goto number;
		case 'w':
			width = 1;
			val = timeptr->tm_wday;
			goto number;
		case 'W':
			val = weeknum(timeptr, MON);
			goto number;
		case 'X':
		case 'T':
			tfmt = "%H:%M:%S";
			goto recursive;
		case 'x':
			tfmt = "%m/%d/%y";
			goto recursive;
		case 'y':
			val = timeptr->tm_year%100;
			goto number;
		case 'Y':
			width = 4;
			val = 1900 + timeptr->tm_year;
			goto number;
		case 'z':
			inc = timezone(s, n, timeptr);
			break;
		case 'Z':
			memcpy(s, timeptr->tm_zone, 3);
			inc = 3;
			break;
		case '\0':
			inc = 0;
			--format;
			break;
		case '%':
			val = '%';
		character:
			*s = val;
			inc = 1;
			break;
		number:
			inc = dval(s, n, width, fill, val);
			break;
		recursive:
			inc = strftime(s, n+1, tfmt, timeptr);
			break;
		}
	}
	*s = '\0';

	return maxsize - n;
}
